File: python_conversion/texture.py
"""
Texture module for defining and manipulating textures.
"""

class Texture:
    """
    Describes bumpy dual texture from cell adjacency info.

    Attributes:
        alpha (float): Bump height.
        beta (float): Radius.
        gamma (float): Height at radius.
    """

    def __init__(self, alpha, beta, gamma):
        """
        Initialize a Texture with float values.

        Args:
            alpha (float): Bump height.
            beta (float): Radius.
            gamma (float): Height at radius.
        """
        self.alpha = alpha
        self.beta = beta
        self.gamma = gamma

    @classmethod
    def from_ints(cls, a, b, c):
        """
        Alternative constructor from integer values (0-100 mapped to 0.0-1.0).

        Args:
            a (int): Alpha value (0-100 mapped to -1.0 to 1.0).
            b (int): Beta value (0-100 mapped to 0.01 to 0.99).
            c (int): Gamma value (0-100 mapped to -1.0 to 1.0).

        Returns:
            Texture: A new Texture instance.
        """
        alpha = max(-1.0, min(1.0, a / 100.0))
        beta = max(0.01, min(0.99, b / 100.0))
        gamma = max(-1.0, min(1.0, c / 100.0))
        return cls(alpha, beta, gamma)

# Predefined textures
Texture.FLAT = Texture(0, 10, 0)
Texture.BUMP = Texture(10, 17, 20)
Texture.SPIKE = Texture(100, 60, 20)
Texture.WEB = Texture(-30, 30, -30)
Texture.HAIRY = Texture(100, 90, 0)

File: original_java_code/GrowthForm4/src/Texture.java
//describes bumpy dual texture from cell adjacency info
public class Texture {

   public double alpha;   //bump height
   public double beta;    //radius
   public double gamma;   //height at radius
   
   public static Texture FLAT = new Texture(0, 10, 0);
   public static Texture BUMP = new Texture(10, 17, 20);
   public static Texture SPIKE = new Texture(100, 60, 20);
   public static Texture WEB = new Texture(-30, 30, -30);
   public static Texture HAIRY = new Texture(100, 90, 0);
           
   public Texture(double alpha, double beta, double gamma) {
      this.alpha = alpha;
      this.beta = beta;
      this.gamma = gamma;
   }  

   //Constructor from integers.   0-100 becomes 0.0-1.0
   public Texture(int a, int b, int c) {
      this.alpha = Math.max(-1.0, Math.min(1.0,  a/100.0));
      this.beta  = Math.max(0.01, Math.min(0.99, b/100.0));
      this.gamma = Math.max(-1.0, Math.min(1.0,  c/100.0));
   }

}
